GH-89: (Minimal) System Monitor Plugin (a.k.a. SysMon)

tacc-wbomar opened the pull request:
# Important

The PR https://github.com/TACC/Core-CMS-Resources/pull/67 is dependent on this PR.

# To Do

- [x] 0d9491f (Optional) Rename `sysmon` to `system_monitor`.\*
- [x] a36e978 [Remove unused `get_direction_classname` import.](https://github.com/TACC/Core-CMS/pull/297#discussion_r687169681)
- [x] (no change) [Revisit `setStatusStyle()`.](https://github.com/TACC/Core-CMS/pull/297#discussion_r687187046)
- [x] (no change) [Re-evaluate where to place constants.](https://github.com/TACC/Core-CMS/pull/297#discussion_r687176885)
- [x] 5ed2892, a2807a2 [Resolve unexpected plugin installs (and search for related problems)](https://github.com/TACC/Core-CMS/pull/297#pullrequestreview-731178122)
- [x] 7220bc1 [Ensure multiple instances (a) can exist or (b) are not permitted to exist.†](https://github.com/TACC/Core-CMS/pull/297#pullrequestreview-731178122)
- [x] 333918d Fix warning badge missing background color.

<details>
<summary>Footnotes</summary>

\* Rename to be consistent with [upcoming `system_specs`](https://github.com/TACC/Core-CMS/compare/task/GH-88-system-specs) and possible `system_...`.
† Either [(a) do not use `GetElementById`](https://github.com/TACC/Core-CMS/commit/7220bc1fbe038973bdd85729beea871f17385029) or ~~(b) limit plugin to one instance~~. _Try not to over-engineer the JavaScript, because #295 will delete it._

</details>

# Overview

Implement _minimal_ SysMon plugin using polished CSS and JavaScript.

# Issues

* #89

# Changes

- Create plugin from snippet.

# Screenshots

## Core

<details open>
<summary>Toggle visibility</summary>

### Form
![Sysmon Plugin - Admin Form](https://user-images.githubusercontent.com/62723358/127896948-9eea5379-cbe8-480a-9d59-fd7716c1a25e.png)

### Render
![Sysmon Plugin - Core Styles](https://user-images.githubusercontent.com/62723358/127896950-2ac2a6c4-0528-4939-8468-dc18951b2a16.png)

### Structure
![System Plugin - in Structure](https://user-images.githubusercontent.com/62723358/127896952-cd6f7b23-16e0-4dd1-aafb-6da6f9c600a0.png)

</details>

## Frontera

<details>
<summary>Toggle visibility</summary>

### Render
![Sysmon Plugin - Frontera Home - Render](https://user-images.githubusercontent.com/62723358/127908190-224ee788-0719-4627-9f84-0b0d1356fb95.png)

### Structure
![Sysmon Plugin - Frontera Home - Structure](https://user-images.githubusercontent.com/62723358/127908202-760326b5-7e49-4cbe-b2b1-067305688fa1.png)

</details>

# Testing

> Last recorded deploy to [Frontera (dev)](https://dev.fronteraweb.tacc.utexas.edu/) was [Build #383 (Aug 2, 2021 1:58:12 PM)](https://jenkins01.tacc.utexas.edu/job/Core_Portal_Deploy/383/).

> __Important__: If you tested this plugin before it was renamed from "Sysmon" to "System Monitor", then you must follow these cleanup steps to prevent a CMS server crash:
>
> <details>
> <summary><strong>Steps for Plugin Rename</strong></summary>
>
> 1. Checkout commit or branch that does not crash server.
>     - If you have only been testing this plugin, then a36e978 should work.
>     - If you have been testing several plugins, then `site/frontera/main` might work.
> 2. Delete "Sysmon" plugin instances:
>     ```
>     python3 manage.py cms uninstall plugins TaccsiteSysmonPlugin
>     ```
> 3. Checkout `task/GH-89` branch at latest commit.
> 4. Perform database migrations for new "System Monitor" plugin:
>     ```
>     python3 manage.py migrate
>     ```
> 5. Re-add deleted plugin instances (as necessary for testing) using new "System Monitor" plugin.
> 6. Rebuild assets and collect static files:
>     ```
>     npm run build && docker exec -it core_cms python manage.py collectstatic --no-input
>     ```
>
> </details>

1. If no environment, then [prepare to test CMS changes](https://github.com/TACC/Core-CMS/wiki/Test-CMS-Changes).
2. In CMS container, run `python3 manage.py migrate`.
3. On CMS website, logged in as admin, add "TACC Site" > "System Monitor" plugin.
4. Reload page (to trigger client-side\* data retrieval).
5. Ensure the following:
    - styles match system monitor on [Old Homepage][old-home] / [Phase 0 Homepage][new-home]
        - the text may be larger; this is acceptable†
    - data shown is for system chosen
    - plugin description (next to name) in Structure view is the system chosen
6. Add "Generic" > "Text" plugin.
7. In Text plugin, in "CMS Plugins" dropdown, confirm "TACC Site" > "System Monitor" plugin is **not** available.
8. Add another "TACC Site" > "System Monitor" plugin, and choose "Stampede2".
9. Ensure Stampede2 monitor—
    - if on local server—has "Warning" state
    - if on remote server—is populated with live data

<details>
<summary>Footnotes</summary>

\* I hope to implement server-side retrieval later via #295.
† Different projects and templates have different root font size. ([Values have been standardized](https://confluence.tacc.utexas.edu/x/nh4FDg), but implementation has not caught up to standards.)

</details>

# Notes

This is a _minimal_ version of a functionally better plugin, [bitbucket:rochaa/tacc-sysmon][bb-sysmon]. @tacc-wbomar has deadline to finish #88 (which requires #89) this week. So #295 has been created to enhance this plugin later.

[bb-sysmon]: https://bitbucket.org/rochaa/tacc-sysmon
[old-home]: https://frontera-portal.tacc.utexas.edu/home-before-2021-07/
[new-home]: https://frontera-portal.tacc.utexas.edu/

tacc-wbomar commented on taccsite_cms/contrib/taccsite_sysmon/cms_plugins.py:
```diff
@@ -0,0 +1,53 @@
+class TaccsiteSysmonPlugin(CMSPluginBase):
+    """
+    Components > "System Monitor" Plugin
+    https://url.to/docs/components/sysmon/
```
I have been using this URL consistently (I think… so code reviews are good) for plugins that have no Confluence doc yet (would be within [… CMS UI - Components / Patterns](https://confluence.tacc.utexas.edu/pages/viewpage.action?pageId=117571624)).

<details>
<summary>Details</summary>

Ideally, the URL is an automatically generated instance of the plugin in use on Core CMS, but that feature does not exist, yet (a [manual dev version](https://dev.cep.tacc.utexas.edu/design-system/pattern-library/) is available, and I have a more populated version on my local database).

</details>

tacc-wbomar replied:
__Note:__ I have been using this URL consistently (I think… so code reviews are good) for plugins that have no Confluence doc yet (would be within [… CMS UI - Components / Patterns](https://confluence.tacc.utexas.edu/pages/viewpage.action?pageId=117571624)).

<details>
<summary>Details</summary>

Ideally, the URL is an automatically generated instance of the plugin in use on Core CMS, but that feature does not exist, yet (a [manual dev version](https://dev.cep.tacc.utexas.edu/design-system/pattern-library/) is available, and I have a more populated version on my local database).

</details>

tacc-wbomar commented on taccsite_cms/contrib/taccsite_sysmon/static/taccsite_sysmon/css/sysmon.css:
```diff
@@ -0,0 +1,86 @@
+/*
```
__Note:__ This code is the styles from https://github.com/TACC/Core-CMS-Resources/pull/67/files __plus__ a dockblock, which is deleted in [`taccsite_custom` change](https://github.com/TACC/Core-CMS/pull/297/files#diff-00db495ec685cbdb37dc9dfede576323eb49fd5d4156be453590c5b7650bb9e6).

tacc-wbomar commented on taccsite_cms/contrib/taccsite_sysmon/static/taccsite_sysmon/js/sysmon.js:
```diff
@@ -0,0 +1,168 @@
+// GH-295: Use server-side logic instead of client-side
+
```
__Note:__ This code is a polished version of the script in https://github.com/TACC/Core-CMS-Resources/pull/67/files, which is deleted in [`taccsite_custom` change](https://github.com/TACC/Core-CMS/pull/297/files#diff-00db495ec685cbdb37dc9dfede576323eb49fd5d4156be453590c5b7650bb9e6).

tacc-wbomar commented on taccsite_cms/contrib/taccsite_sysmon/templates/sysmon.html:
```diff
@@ -0,0 +1,34 @@
+{% load static %}
+
```
__Note:__ This code is mostly the markup in https://github.com/TACC/Core-CMS-Resources/pull/67/files, which is deleted in [`taccsite_custom` change](https://github.com/TACC/Core-CMS/pull/297/files#diff-00db495ec685cbdb37dc9dfede576323eb49fd5d4156be453590c5b7650bb9e6).

tacc-wbomar reviewed (commented):
Added notes for other reviewers.

duckonomy commented on taccsite_cms/contrib/taccsite_sysmon/cms_plugins.py:
```diff
@@ -0,0 +1,53 @@
+from django.utils.translation import gettext_lazy as _
+
+from taccsite_cms.contrib.helpers import concat_classnames
+from taccsite_cms.contrib.taccsite_offset.cms_plugins import get_direction_classname
```
Where/what is this used for?

tacc-wbomar replied:
Deleted via a36e978.

duckonomy commented on taccsite_cms/contrib/taccsite_sysmon/models.py:
```diff
@@ -0,0 +1,45 @@
+from .constants import DEFAULT_SYSTEM
+
+# TODO: (Maybe in GH-295) Do not replicate `display_name` data from API
+SYSTEM_DICT = {
```
Is there a reason that this isn't in `constants.py`?

tacc-wbomar replied:
I don't mind moving them (for this PR, and the other); `constants.py` is where I would naturally put them. Anyone else suggesting to do so is the only catalyst I need to do so.

tacc-wbomar replied:
I refrain from moving constants, because:
1. [your "valid reason" comment/approval](https://github.com/TACC/Core-CMS/pull/297#issuecomment-897161469)
2. [my original reasoning](https://github.com/TACC/Core-CMS/pull/297#discussion_r687176885)
3. consistency with other recent plugins

duckonomy commented on taccsite_cms/contrib/taccsite_sysmon/static/taccsite_sysmon/js/sysmon.js:
```diff
@@ -0,0 +1,168 @@
+ */
+function setStatusStyle(type) {
+  switch (type) {
+    case 'warning':
```
Will there be more `type`s to compare against? If not, is there reason that `switch` used here (in comparison to other blocks?

tacc-wbomar replied:
I began the work for this, but then abandoned it, becuase it grew beyond the scope of a "minimal" plugin. I have added a requirement on #295 to support multiple, distinct states (rather than assume failure and ignore loading).

duckonomy reviewed (approved):
I've approved! However, I also included some minor questions/comments.

tacc-wbomar reviewed (commented):
@duckonomy, thanks. I've responded.

Also, what you think of renaming this plugin (to mirror [`system_specs` plugin](https://github.com/TACC/Core-CMS/compare/task/GH-88-system-specs) name)?

<img width="556" alt="Screen Shot 2021-08-11 at 16 02 46" src="https://user-images.githubusercontent.com/62723358/129103132-46fd351b-ee81-4d40-bb41-243f357810b3.png">

duckonomy commented:
> @duckonomy, thanks. I've responded.
> 
> Also, what you think of renaming this plugin (to mirror [`system_specs` plugin](https://github.com/TACC/Core-CMS/compare/task/GH-88-system-specs) name)?
> 
> <img alt="Screen Shot 2021-08-11 at 16 02 46" width="556" src="https://user-images.githubusercontent.com/62723358/129103132-46fd351b-ee81-4d40-bb41-243f357810b3.png">

That sounds good!

duckonomy commented:
@tacc-wbomar For 2,3 I think those are valid reasons 👍🏼

tacc-wbomar commented:
@duckonomy I have re-requested review merely to remove approval to help remind me that this PR is **not** complete. There is nothing new to review, yet.

tacc-wbomar commented:
@duckonomy I completed the To Do items. Code was changed only for:

- removal of errant import (a36e978)
- plugin rename (0d9491f)
    - To test, follow "Steps for Plugin Rename" in "Testing" section of description.
- addition of missing functions (871676e)
    - I found this bug during my re-test.

Please review.

duckonomy commented on taccsite_cms/settings.py:
```diff
@@ -241,6 +241,11 @@ def getsecrets():
     # TODO: Extract TACC CMS UI components into pip-installable plugins
     # FAQ: The djangocms_bootstrap4 library can serve as an example
     'taccsite_cms.contrib.taccsite_sample',
+    'taccsite_cms.contrib.taccsite_static_article_list',
+    'taccsite_cms.contrib.taccsite_static_article_preview',
+    'taccsite_cms.contrib.taccsite_blockquote',
+    'taccsite_cms.contrib.taccsite_offset',
```
These cause an error on initialization (I had to comment them to make it run properly).

tacc-wbomar replied:
That is a problem. Three of those should not be there. This problem suggests that more code may be present that should not be. I will investigate. Thank you.

<details>
<summary>Details</summary>

_There are so many dependent branches [for Frontera System Specs page](https://github.com/TACC/Core-CMS/pull/313) and [Frontera Homepage plugins](https://github.com/TACC/Core-CMS/pull/259) that I am trying to author independently but test together 🤕 and I crossed the streams._

</details>

tacc-wbomar replied:
I have removed those lines (via 5ed2892). No related erroneous code was found, but I found and resolved\* a different unexpected change in the submodule diff.

<details>
<summary>Footnotes</summary>

\* I resolved unexpected submod diff with a merge of `main`, a2807a2 (after a submod pointer update: #315).

</details>

tacc-wbomar commented:
I have added a task to either support multiple instances or not.

<details>
<summary>Details</summary>

_The problem is caused by the JavaScript I borrowed and cleaned up still using `document.getElementById`. There really should be only one instance of this plugin on a page. I need to either (with Python) prevent multiple instances __or__ (with JavaScript) allow multiple instances._

</details>

tacc-wbomar commented:
@duckonomy, fixes:

- Multiple instances are now functional.
    <details><summary>For just the diff relevant to this…</summary>

	_I suggest reviewing first 30c78a6, then 7220bc1 (instead of "changes since last review")._

    </details>

- Also, the warning state is now testable.
    <details><summary>For just the diff relevant to this…</summary>

	_I suggest reviewing 333918d (instead of "changes since last review")._

    </details>

duckonomy reviewed (approved):
LGTM!
